src/riscv/linux/cpuinfo.c

```diff
@@ -0,0 +1,315 @@
+        }
+    }
+    /* Skip line if no ':' separator was found. */
+    if (separator == line_end) {
+        cpuinfo_log_info("Line %.*s in /proc/cpuinfo is ignored: key/value separator ':' not found",
+                         (int) (line_end - line_start), line_start);
+        return true;
+    }
+
+    /* Skip trailing spaces in key part. */
+    const char* key_end = separator;
+    for (; key_end != line_start; key_end--) {
+        if (key_end[-1] != ' ' && key_end[-1] != '\t') {
+            break;
+        }
+    }
+    /* Skip line if key contains nothing but spaces. */
+    if (key_end == line_start) {
+        cpuinfo_log_info("Line %.*s in /proc/cpuinfo is ignored: key contains only spaces",
+                         (int) (line_end - line_start), line_start);
+        return true;
+    }
+
+    /* Skip leading spaces in value part. */
+    const char* value_start = separator + 1;
+    for (; value_start != line_end; value_start++) {
+        if (*value_start != ' ') {
+            break;
+        }
+    }
+    /* Value part contains nothing but spaces. Skip line. */
+    if (value_start == line_end) {
+        cpuinfo_log_info("Line %.*s in /proc/cpuinfo is ignored: value contains only spaces",
+                         (int) (line_end - line_start), line_start);
+        return true;
+    }
+
+    /* Skip trailing spaces in value part (if any) */
+    const char* value_end = line_end;
+    for (; value_end != value_start; value_end--) {
+        if (value_end[-1] != ' ') {
+            break;
+        }
+    }
+
+    const uint32_t processor_index      = state->processor_index;
+    const uint32_t max_processors_count = state->max_processors_count;
+    struct cpuinfo_riscv_linux_processor* processors = state->processors;
+    struct cpuinfo_riscv_linux_processor* processor  = &state->dummy_processor;
+    if (processor_index < max_processors_count) {
+        processor = &processors[processor_index];
+    }
+
+    const size_t key_length = key_end - line_start;
+    switch (key_length) {
+        case 3:
+            if (memcmp(line_start, "mmu", key_length) == 0) {
+                // Do nothing
+            } else if (memcmp(line_start, "isa", key_length) == 0) {
+                parse_isa(value_start, value_end, processor);
+            } else {
+                goto unknown;
+            }
+            break;
+        case 4:
+            if (memcmp(line_start, "hart", key_length) == 0) {
+                // Do nothing
+            } else {
+                goto unknown;
+            }
+            break;
+        case 5:
+            if (memcmp(line_start, "uarch", key_length) == 0) {
+                parse_uarch(value_start, value_end, processor);
+            } else {
+                goto unknown;
+            }
+            break;
+        case 9:
+            if (memcmp(line_start, "processor", key_length) == 0) {
+                const uint32_t new_processor_index = parse_processor_number(value_start, value_end);
```

Comment:
In the case where we continuously fail to read the line and always return 0, I think these conditions will just repeatedly overwrite the state of the '0' processor? We need to also have a condition here where the processor index never progresses. (Or hard-fail when 'parse_processor_number' fails).